### ISSUE TITLE
🧊 [Feature] Support FBX Model Animations.

### ISSUE THREAD

**Issue by MrScautHD** [MonoGame/MonoGame#7976]
Hey!
I would love if you can Support Model Animations!
like a Method that you can call in the Model class (i hope i come in :))

model.StartAnimation(1);

#### What version of MonoGame does the bug occur on:
- MonoGame 3.8.1

#### What operating system are you using:
- Windows

#### What MonoGame platform are you using:
Cross Platform

**Comment by ohtrobinson** [comment]
This is getting into game engine territory, and there are already a few libraries that do exactly this.

**Comment by MrScautHD** [comment]
...

> This is getting into game engine territory, and there are already a few libraries that do exactly this.

No, thats for real no reason

**Comment by Beyley** [comment]
animations and rigging is something better handled by a game engine (which monogame is not), there are already libraries that will do a 10x better job specializing in this than monogame ever could

**Comment by MrScautHD** [comment]
> animations and rigging is something better handled by a game engine (which monogame is not), there are already libraries that will do a 10x better job specializing in this than monogame ever could

For real it looks always more that MonoGame is a very bad lib that can do nothing good.

I mean the code is a mess...
It Support mostly nothing or just old and messy stuff :(

**Comment by MrScautHD** [comment]
MonoGame try to copy XNA but who say that is a good idea...?

I mean in the only thing in that MonoGame is good is that it has the right concept but, if we see the quality of MonoGame would it better to switch to anything better

**Comment by MrScautHD** [comment]
I mean why you guys not try to improve MonoGame, you always say XNA dont't has it so MonoGame should it not have, i think the idea of MonoGame should go to MonoGame should easy to use and add features to go amazing.

For real i love the idea of the Framework no disgusting code generation or editors like Unity but you guys not try to improve it just to copy things, and own features!

But if everyone just say mhh no MonoGame would never add this so close this suggest is just bad and i think thats the reason why MonoGame is going in a wrong way

**Comment by vpenades** [comment]
Something I've criticised in the past is that MonoGame tries to sell itself as a competitor to Unity and other game engines that provide a full solution to game development. As a consecuence, people like you @MrScautHD  end with the misconception that MonoGame must do the same.

MonoGame is a foundation, to build on top of it the exact engine you need, so by itself if offers nearly nothing. You need to combine MonoGame with other building blocks to have something that looks like a game engine. And yes, in some cases that means that you'll need to write some of the building blocks by yourself. But that has the advantage of having building blocks tailored for your needs.

**Comment by MrScautHD** [comment]
I can understand that, BUT why support MonoGame models if you need to write you own Model system Because it not support Animations, i not even would extend the Model class i mean did you ever looked on this mess?

**Comment by MrScautHD** [comment]
If you do things that do it perfect!

**Comment by MrScautHD** [comment]
and not 50% / 100%

**Comment by Ravendarke8** [comment]
Alright, let's pull full break on this one. Only true issue with monogame is their mod team giving breathing space to people like you. You are absolute beginner, you can't recognize messy or well written code. It took you 5 days to write few lines code for what you call "3D camera" after you begged for two days for someone else to do it basically for you. Result is still horrible, gimbal lock is probably swear word to you.

You expect every little thing being handed to you, every basic, camera handling, model handling, animation, imports, yet you refuse to use engine which were made EXACTLY for people like you. Instead you are picked framework and do one of two things:
1) beg for library to do something for you
2) have someone implement specific engine function into existing framework

It was kinda fun to watch for a while but as of the moment it turns into annoying spam. You talk about implementing full model rigging system but at same time you literally do not understand face winding, after two ppl spend hour explaining it to you. You are hard stuck on incredibly high Dunn Kruger peak and there is no movement in positive direction. Either go back to using engine of your choice, or start learning basics and build your own, using whatever framework you want.

**Comment by vpenades** [comment]
> i mean did you ever looked on this mess?

Yes, of course I did, we all did, this is legacy code, but there's something called "backwards compatibility" that might not be important to you, but it's important to other developers.  You're not the only one using monogame.

Also notice that "this mess" the `Model ` class is, was inherited by MonoGame from the original XNA code written by Microsoft, if you want to blame someone about bad code design, talk to MS.

> If you do things then do it perfect!

Are you paying for it?  Keep in mind that people that contributes to MonoGame, DOES IT FOR FREE, coming here demanding things is not the best way of getting things done.

I am the first one that does not like many things about MonoGame, but instead of blaming everybody for not working for me for free, I actually contribute with bug fixes and some of the building blocks.

So, you want to load animated models into MonoGame? here we go: [MonoScene](https://github.com/vpenades/MonoScene)

**Comment by MrScautHD** [comment]
> Alright, let's pull full break on this one. Only true issue with monogame is their mod team giving breathing space to people like you. You are absolute beginner, you can't recognize messy or well written code. It took you 5 days to write few lines code for what you call "3D camera" after you begged for two days for someone else to do it basically for you. Result is still horrible, gimbal lock is probably swear word to you.
> 
> You expect every little thing being handed to you, every basic, camera handling, model handling, animation, imports, yet you refuse to use engine which were made EXACTLY for people like you. Instead you've picked framework and do one of two things:
> 
>     1. beg for library to do something for you
> 
>     2. have someone implement specific engine function into existing framework
> 
> 
> It was kinda fun to watch for a while but as of the moment it turns into annoying spam. You talk about implementing full model rigging system but at same time you literally do not understand face winding, after two ppl spend hour explaining it to you. You are hard stuck on incredibly high Dunn Kruger peak and there is no movement in positive direction. Either go back to using engine of your choice, or start learning basics and build your own, using whatever framework you want.

full false

**Comment by MrScautHD** [comment]
What ever i not wanna start with you guys a war, but to say to every suggest no it just unmotivated

**Comment by MrScautHD** [comment]
And sorry for being so disrespectful by saying MonoGame is a mess...

**Comment by ohtrobinson** [comment]
If you think the model class is so bad, fix it yourself.

Is it the best thing in the world? No.
Should we expect it to be fixed by the maintainers in their own time for free when if we think it is so bad we could just fix it ourself? No.

That's the thing about open source. if you look at something and think "that's horrible, it could be done better by doing xyz", then do it yourself and submit a PR. if the maintainers like it they'll have no issues merging it in so that you, and many more users can enjoy the benefits.

Granted I doubt that will happen in this case because it's based off the XNA model stuff and I doubt they would want to deviate from that just yet.

**Comment by MrScautHD** [comment]
I made this suggest for do a own Pull to see if it would get added or not. (i not wanna waste my time if it get not merged)

**Comment by MrScautHD** [comment]
But to say MonoGame would never add it with the reason MonoGame should not go to a Game Engine... is just stupid

**Comment by Apostolique** [comment]
It's more likely that those engine-wanna-be features would get axed in the future instead of being developed more. The only main new feature I expect MonoGame to get is IMGUI even though I personally would be against it. A lot of cruft should get removed from MonoGame and be split into separate packages a bit like Katabasis is already doing.

**Comment by MrScautHD** [comment]
i understand that you want MonoGame should support base things like

Rendering, Tick, Sounds, Fonts... 

But nothing is wrong with few more features

**Comment by Beyley** [comment]
> i understand that you want MonoGame should support base things like
> 
> Rendering, Tick, Sounds, Fonts...
> 
> But nothing is wrong with few more features

As has been explained multiple times in your various issues, there are definitely things wrong with more features, especially when the features do not line up with the goals of the project, and are likely to be left to rot, then removed at a later date

**Comment by stromkos** [comment]
> i understand that you want MonoGame should support base things like
> 
> Rendering, Tick, Sounds, Fonts...
> 
> But nothing is wrong with few more features

There is nothing wrong with new, non-breaking features.

 I have a wish list of my own.   
Most of which are  breaking changes from XNA.

Is it a good idea, sure, but you have burned the goodwill from the community with your other posts,.

Having worked with FBX projects at both the raw level and the AutoDesk implementation,  what you are asking is not trivial. 

Yes, the system is flawed,  but it is the best we have, and I support  it.

No system is perfect, but this one tends to work well.

**Comment by MrScautHD** [comment]
> 

yea its a point, but i just want to say features should not instandly denied lets discutate about it.

**Comment by Xrayez** [comment]
> As has been explained multiple times in your various issues, there are definitely things wrong with more features, especially when the features do not line up with the **goals** [emphasis mine] of the project, and are likely to be left to rot, then removed at a later date

@Beyley, while is true, I do not see any goals explicitly listed in MonoGame's documentation. The only thing I see documented is the fact that "*MonoGame is a re-implementation of the discontinued Microsoft's XNA Framework*". Other things that can be documented include: priorities, non-goals, scope, etc.

So I'd definitely suggest to @MrScautHD not to waste the time on these feature proposals for MonoGame, whatever the intentions. The only thing which I'd legimitely request for is documentation for claimed MonoGame's goals, something which @vpenades already stated:

> Something I've criticised in the past is that **MonoGame tries to sell itself as a competitor to Unity** [emphasis mine] and other game engines that provide a full solution to game development. As a consecuence, people like you @MrScautHD end with the misconception that MonoGame must do the same.

If both new and veteran users think or used to think that way, then I do think MonoGame project does have a documentation issue in this regard. New users may no longer think that MonoGame attempts to be comparable to Unity, of course, but new users may still expect that new features are going to be added to MonoGame, which may not be the actual case.

**Comment by vpenades** [comment]
To clarify my statement about "MonoGame tries to sell itself as a competitor to Unity" it's not MonoGame's official position, but the fact that, whenever a newcomer asks either here or in the forums if MonoGame can do _this_ or _that_, the average answer he receives is "definitely yes" but most of the time without specifying that such feature X is provided by a third party solution and not by MonoGame itself.

Personally, I think that MonoGame should be as bare bones as possible, and let most of the features to be implemented by third parties.

**Comment by Xrayez** [comment]
Yes, I mean if both you and the OP had this impression in the past, then it's a mismatch of expectations vs reality.

> without specifying that such feature X is provided by a third party solution and not by MonoGame itself.

This is something I've noticed with other open-source projects, as if third-party solutions are part of an upstream/original project, regardless the scope and scale of a project. I think this may have something to do with overzealous fans that may see everything associated with the original project as part of the unified ecosystem.

On the other hand, MonoGame also seems to presents itself as the "one" framework:

![image](https://user-images.githubusercontent.com/17108460/211160177-f7f0080f-de45-4ebe-8120-5922e405ab49.png)

The wording matters! 😉

**Comment by squarebananas** [comment]
Some related information regarding FBX animations which may be of interest.

A blog post by Shawn Hargreaves (one of the original XNA developers) detailing the reasoning behind the way Models are implemented:
https://shawnhargreaves.com/blog/note-to-self-model-hammer.html

The archived Animation sample from the official Microsoft XNA site:
https://github.com/SimonDarksideJ/XNAGameStudio/tree/archive/Samples/CustomModelAnimation_4_0

Personally I have referred to the above sample before although I use my own code for combining animation files, instancing, blended animation switching, etc. Having the animation data is only one part of an animation system which may differ greatly between projects.

**Comment by aurodev** [comment]
This "thread", which it shouldn't be, is exactly what you get when you feed trolls.
